### PR TITLE
Added UriInfo#getMatchedResourceTemplate method

### PR DESCRIPTION
**Pull request by jansupol**
Implements #1191 

There can be a discussion of the actual method name, as the proposed name was `getResourceTemplate`, but `getMatchedResourceTemplate` is more aligned with other `matched` methods.

Also, the return type can be in the form of a list, but the discussion was to return a string

### CODE REVIEW

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/UriInfo.java [hunk 1]
```diff
@@ -230,6 +231,70 @@ public interface UriInfo {
      */
     public List<String> getMatchedURIs();
 
+    /**
+     * <p>
+     * Get a template concatenated of all {@link Path Paths} matched by all the resources when matching the current request.
+     * </p>
+     * <p>
+     * Each {@link Path} value used to match a resource class, a sub-resource method or a sub-resource locator is concatenated
+     * to a single {@code String} value. The template does not include query parameters but do include matrix parameters
```
Comment (jim-krueger) [inline jaxrs-api/src/main/java/jakarta/ws/rs/core/UriInfo.java:240]: ...but does include

**Review by jim-krueger — CHANGES_REQUESTED** [review]
Should there be TCK additions to ee.jakarta.tck.ws.rs.ee.rs.core.uriinfo.URIInfoTest?

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/UriInfo.java (hunk 1 above)
Comment (spericas) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/UriInfo.java:240]: "to a single" -> "into a single"?

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/UriInfo.java [hunk 2]
```diff
@@ -230,6 +231,70 @@ public interface UriInfo {
      */
     public List<String> getMatchedURIs();
 
+    /**
+     * <p>
+     * Get a template concatenated of all {@link Path Paths} matched by all the resources when matching the current request.
```
Comment (spericas) [inline jaxrs-api/src/main/java/jakarta/ws/rs/core/UriInfo.java:236]: Suggestion: "Get a URI template that includes all paths matched by the current request's URI."

**Comment by jansupol** [comment]
> Should there be TCK additions to ee.jakarta.tck.ws.rs.ee.rs.core.uriinfo.URIInfoTest?

Definitely. I plan to add those if accepted

**Comment by jim-krueger** [comment]
Would be nice to get this into 4.0.  Not sure if there is enough runway.

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/UriInfo.java (hunk 1 above)
Comment (jansupol) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/UriInfo.java:240]: @jim-krueger what do you mean by "does include"? The URI template itself does not contain the query parameter in the request unless the URI template contains a string containing the `\\?` symbol.

Comment (jim-krueger) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/UriInfo.java:240]: Sorry, should have been more clear.   
`but do include matrix parameters`
should be:
`but does include matrix parameters`

**Comment by jansupol** [comment]
Added the TCK tests

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs40/ee/rs/core/uriinfo/JAXRSClientIT.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import java.io.IOException;
+
+/*
+ * @class.setup_props: webServerHost;
+ *                     webServerPort;
+ */
+@ExtendWith(ArquillianExtension.class)
+public class JAXRSClientIT extends JAXRSCommonClient {
```
Comment (jamezp) [inline jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs40/ee/rs/core/uriinfo/JAXRSClientIT.java:38]: Minor nit, but could we name new TCK tests something more relatable? It's confusing when you just get a report of "JAXRSClientIT" tests failing.

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs40/ee/rs/core/uriinfo/JAXRSClientIT.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+  private static final long serialVersionUID = 40L;
```
Comment (jamezp) [inline jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs40/ee/rs/core/uriinfo/JAXRSClientIT.java:40]: Again, minor nit, but I don't think there is any need for this.

**Comment by jim-krueger** [comment]
With the TCK I am in favor of adding this to the Jakarta Rest 4.0 content.   Can we get a quick vote on this?

@ivargrimstad Can we still add this relatively minor, but nice to have,  improvement to 4.0 (assuming it passes a vote)?

**Review by jim-krueger — APPROVED** [review]

**Comment by jansupol** [comment]
An alternative with returning the List of templates #1245

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs40/ee/rs/core/uriinfo/UriInfoTestResource.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+@Path("/resource")
+public class UriInfoTestResource {
+    public static final String ONE_POST = "{name:[a-zA-Z][a-zA-Z_0-9]*}";
+    public static final String TWO_GET = "{Prefix}{p:/?}{id: ((\\d+)?)}/abc{p2:/?}{number: (([A-Za-z0-9]*)?)}";
+    public static final String TWO_POST = "{Prefix}{p:/?}{id: ((\\d+)?)}/abc/{yeah}";
+    public static final String THREE_SUB = "{x:[a-z]}";
+
+    public static class SubGet {
+        @PUT
+        public String get(@Context UriInfo uriInfo) {
```
Comment (jim-krueger) [inline jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs40/ee/rs/core/uriinfo/UriInfoTestResource.java:36]: Unless I'm mistaken, in order to fully test what Andrew requested this method would need a `@Path(/something)`

**Review by jim-krueger — APPROVED** [review]

**Review by spericas — APPROVED** [review]

**Comment by jansupol** [comment]
I slightly prefer the #1245 over this one

**Review by jamezp — APPROVED** [review]
In talking with some of the OTel folks, they seemed to prefer the string return type so they don't have to combine the list to a string.

**Pull request merged**